cpp/src/tests/sort/sort_test.cu

```diff
@@ -0,0 +1,443 @@
+// -*-c++-*-
+
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+#include "cuda_profiler_api.h"
+
+#include "sort/sort.cuh"
+#include "rmm_utils.h"
+#include "test_utils.h"
+
+#include <chrono>
+
+#include <curand_kernel.h>
+
+#define MAX_NUM_GPUS 16
```

Comment:
I am not an assigned reviewer, but seeing the email thread on this PR, this worries me. Is this future proof? What happens if NVIDIA releases a new system with more than 16 GPUs? (Or if another company custom builds a system with more than 16 GPUs in a single node?) A bug due to hard-coding MAX_NUM_GPUS happened in hornet, and this can happen here as well in the future. I think we should better query the number of GPUs in the target system than hard-coding MAX_NUM_GPUS.

Reply:
Or we should at least add a check routine comparing the queried number of GPUs and MAX_NUM_GPUS and if # GPUs > MAX_NUM_GPUS, abort the execution with error message to avoid undefined behaviors which can be hard to debug.

Reply:
The number of GPUs is set a template parameter for the class and can be changed based on needs. This is different than Hornet that was defining it as a macro (which it shouldn't have...)

Reply:
@ogreen yes, it's a template parameter elsewhere, but in this file, I see "#define MAX_NUM_GPUS 16" (sort_test.cu line 32).

and later (sort_test.cu, from line 217), this can lead to writing passing the array boundary.

```
  for (int i = 0 ; i < n_gpus ; ++i) {
    cudaSetDevice(i);

    EXPECT_EQ(RMM_FREE(d_input[i], stream), RMM_SUCCESS);
    EXPECT_EQ(RMM_FREE(d_output[i], stream), RMM_SUCCESS);
  }
```

Reply:
This is just the tester file. We need to define a maximum for testing purpose.

Reply:
I think using std::vector and resizing std::vector based on n_gpus is a safer coding practice, but this is just a recommendation. But I think we should at least check n_gpus <= MAX_NUM_GPUS to avoid surprising undefined behaviors; and sort_impl.cuh is doing this check as well.

Reply:
I have added a check in the test to ensure that the number of opus on the system <= MAX_NUM_GPUS.  Will add more commentary below.

Reply:
I agree that using std::vector would be a safer coding practice.  I actually had a different solution in mind - but felt like I had done enough refactoring of the code thus far to go ahead and get something into the code base - so I made MAX_NUM_GPUS a template parameter for now and used the fixed array sizes.

I can add a refactoring of this logic to the backlog and try and get to that later.